### PR TITLE
chore(index): idiomatise various iterators

### PR DESCRIPTION
**Pull request by passcod**

### CODE REVIEW

**Comment by passcod** [comment]
Feel free to ignore; I just went through index.rs and tried to make the iterators a little nicer. Highlights/summary:

- implementing `Hash` on `SerializableEntry` ourselves so we can use a hashset instead of a hashmap we ignore the keys of. That also lets us collapse the map insertion loop into a `collect()` call.
- writing directly to the bucket in `insert` is a bit more verbose but avoids a string allocation.
- various instances of using better chaining methods, like `filter_map` and more chainy Option/Result usage
- reading files line by line with `BufReader` instead of reading the entire file in a string in `bucket_entries()`

**Comment by zkat** [comment]
omg I'm sorry I completely missed this PR when you put it up. I'm going to try and resolve the conflict myself. I'm actually watching this repo now, too, so I should get notified of anything.

Thank you so much for going through it and helping make it more idiomatic! Lots to learn 💃

**Comment by passcod** [comment]
Haha, no worries! Also a better way to write lines 76-9 (on my side of the diff) would be the `write!` macro, which I myself often forget about!

**Comment by zkat** [comment]
Um, I tried resolving the conflicts and it was a lot. Would it be too much to ask you to do the resolution yourself? 😅

**Comment by passcod** [comment]
Yeah, sure. Might be an hour or so, need to get home first ;)

**Comment by zkat** [comment]
oh totally. Take your time! and thank you again :)

**Comment by passcod** [comment]
Actually while I'm at it I'm going to step through some of this stuff to make it clearer what/why I'm doing it, cause that's probably a little more helpful than a bunch of changes.

### `write!` instead of `format!`-then-write

You had something like this:

```rust
let str = format!("\n{}\t{}", blih, bloh);
OpenOptions::create(true).append(true).open(&path)?.write_all(&str.into_bytes())?;
```

Firstly, as a by-the-by, naming a variable `str`, while not invalid, it kinda super confusing because of `str`-the-type.

Secondly, what's happening here is you're allocating a String, writing your data into it, then converting the string into bytes, and writing that to a file.

What you can instead do is write to the file:

```rust
let mut file = OpenOptions::create(true).append(true).open(&path)?;
write!(&mut file "\n{}\t{}", blih, bloh)?;
```

This works because `File` implements `io::Write`, and `write!` is a macro that writes a format string directly to a buffer that implements `io::Write`.

The way I did it originally in the PR is kinda the long way of the same concept: open up the file, write to it directly.

```rust
file.write_all(b"\n")?;
file.write_all(hash_entry(&stringified).as_bytes())?;
file.write_all(b"\t")?;
file.write_all(stringified.as_bytes())?;
```

### Try not to unwrap inside an unwrap

This is this thing:

```rust
let value: Option<Foo> = ...;
value.unwrap_or_else(|| another_option().unwrap());
```

In that case, you're unwrapping twice! You can chain with `or_else` instead and only unwrap once:

```rust
value.or_else(|| another_option()).unwrap();
```

In your particular case, the inner thing was actually returning a `Result`. That's still ok, because you can transform it into an `Option` using `.ok()`:

```rust
value.or_else(|| another_result().ok()).unwrap();
```

#### Out-of-scope addendum: try not to unwrap, especially when you're already returning a `Result`

This is all happening inside a function with this signature:

```rust
pub fn insert(cache: &Path, key: &str, opts: PutOpts) -> Result<Integrity, Error>
```

So really, you could instead transform the `Option` into an appropriate error.

#### Out-of-scope addendum: almost-static `or_else`

Clippy will warn you about static `or_else` usage, e.g. `or_else(|| 42)`, because that can just be written `.or(42)` without the closure. It won't warn you (yet?) about near-static usage, like you have:

```rust
.or_else(|| "sha1-deadbeef".parse::<Integrity>().ok())
```

This would be an actual constant with a little more preprocessing. As it is, you could probably use lazy_static to evaluate it once, and then use `.or(DEADBEEF)` or something.

### Returning results instead of trying and returning

More of a nit, this is when you "try" (using the `?` operator) something then immediately return a static `Ok(())`:

```rust
fn delete(args: SomeArg) -> Result<(), Error> {
	do_something(args)?;
	Ok(())
}
```

Map it instead:

```rust
fn delete(args: SomeArg) -> Result<(), Error> {
	do_something(args).map(|_| ())
}
```

### Path join instead of PathBuf push

`PathBuf` derefs to `Path`, which has `.join()`, which returns a `PathBuf` (and does not mutate).

So instead of:

```rust
let mut path = PathBuf::new();
path.push(existing_path);
path.push("suffix");
```

You can just do this:

```rust
let path = existing_path.join("suffix");
```

### BufReader is your friend

BufReader implements some additional Reader things that deal with chunks of data instead of going byte by byte. One of the common use cases is reading data line by line.

```rust
use std::io::{BufRead, BufReader};
BufReader::new(file).lines() // <== an iterator
```

#### Call-out: returning early for fun and profit

This may be a bit hard to follow in the diff because the entire function was turned upside down and shaken up, but at the start of yours, there was this:

```rust
let lines = match fs::read_to_string(bucket) {
    Ok(data) => Ok(data),
    Err(ref e) if e.kind() == ErrorKind::NotFound => Ok(String::from("")),
    err => err,
}?;
````

Now, the `Ok(String::from(""))` here (which could be `"".into()`, btw) is ultimately going to be transformed to an empty Vec, by way of trying to split an empty string and then folding an empty iterator into a Vec.

Now, you probably have it that way because this entire expression needs to return an `io::Result<String>`. You could instead write an early return and bypass that...

```rust
let lines = match fs::read_to_string(bucket) {
    Ok(data) => Ok(data),
    Err(ref e) if e.kind() == ErrorKind::NotFound => return Ok(Vec::new()),
    err => err,
}?;
```

But if you look carefully, you'll see that this is another example of the try-then-return thing above:

```rust
let lines = match fs::read_to_string(bucket) {
    Ok(data) => Ok(data),
    Err(ref e) if e.kind() == ErrorKind::NotFound => return Ok(Vec::new()),
    err => err,
}?; // try

Ok(...) // return
```

So you could go further and return early in the err branch as well:

```rust
let lines = match fs::read_to_string(bucket) {
    Ok(data) => Ok(data),
    Err(ref e) if e.kind() == ErrorKind::NotFound => return Ok(Vec::new()),
    err => return err,
}?;
```

And here, there's not much point having two early returns, just consolidate into one:

```rust
let lines = match fs::read_to_string(bucket) {
    Ok(data) => Ok(data),
	Err(e) => return if e.kind() == ErrorKind::NotFound {
		Ok(Vec::new())
    } else {
        Err(e)
    }
}?;
```

And you can drop the try because one branch is always returning, and thus the other never needs to wrap into a Result and then immediately unwrap it again:

```rust
let lines = match fs::read_to_string(bucket) {
    Ok(data) => data,
	Err(e) => return if e.kind() == ErrorKind::NotFound {
		Ok(Vec::new())
    } else {
        Err(e)
    }
};
```

Instead of keeping it like that, I went the chaining route with an `.or_else` at the bottom, but this would have been a pretty valid form, too. The key take-away here is that with early returns, not only can the logic be clearer, but you can escape from strict typing a little, and that becomes extremely worthwhile.

### The big controversive one, split into a second commit for clarity and undo-ability: exploiting the `Hash` trait (on a private struct)

So, this is less of a beautiful idiomatic refactoring and veering slightly more into the not-completely-orthodox. I believe strongly that it is completely acceptable in internals, as an implementation detail, but probably shouldn't be exposed via a public API to avoid surprises.

So, the original code—

Ok, wait, before that, two small things:

- `core::iter::empty().collect::<Vec<Entry>>()` is really just a fancy confusing way of saying `Vec::new()`.
- If you're finding yourself doing a `filter` and then a `map`, use a `filter_map` instead.

Alright, so, the... almost original code:

```rust
let bucket = bucket?;
if bucket.file_type().is_dir() {
    return Ok(Vec::new());
}
let entries = bucket_entries(bucket.path())?;
let mut dedupe: HashMap<String, SerializableEntry> = HashMap::new();
for entry in entries {
    dedupe.insert(entry.key.clone(), entry);
}
let iter = dedupe
    .into_iter()
    .filter_map(|se| if let Some(i) = se.integrity {
        let se = se.1;
        Some(Entry {
            key: se.key,
            integrity: i.parse().unwrap(),
            time: se.time,
            size: se.size,
            metadata: se.metadata,
        })
    } else {
        None
    });
Ok(iter.collect::<Vec<Entry>>())
```

The critical part here is the `HashMap` deduping step. You're using the key of the entry as the deduplicator, pushing into a map, and then iterating it out. Clever. Ultimately, though, what if you could tell the hashmap directly that you're trying to use `entry.key` as a deduper, and also what if you could convert the entire `bucket_entries()` into that map in one step? That would be pretty wonderful.

Well, you can!

You just need some boilerplate.

Instead of using a `HashMap`, we'll use a `HashSet`. The advantage here is that a map is a key-value collection, which is convertible only from a two-tuple, while a set is a value-only collection, which you can convert to easily from any other such value-only collections, like a vec or any kind of iterator.

And to instruct the `HashSet` to use `entry.key` as the deduper, we have to implement `Hash` on your `SerializableEntry`, considering only the `key`. Fortunately, that's pretty easy:

```rust
impl Hash for SerializableEntry {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.key.hash(state);
    }
}
```

Because the set is going to do comparisons, `PartialEq` and `Eq` are also needed. Implemented in the same fashion:

```rust
impl PartialEq for SerializableEntry {
    fn eq(&self, other: &Self) -> bool {
        self.key == other.key
    }
}

impl Eq for SerializableEntry {}
```

These implementations are why this should be an internal technique only: it would be pretty surprising to find that `SerializableEntry`, if it was part of the public API, didn't Hash or Eq according to its full contents, but only to its `key`. There's ways to get around that if you need to, e.g. by creating an internal-only tuple struct around the public type, and implementing the odd `Hash` and `Eq` on that instead, but here's it's all internal so there's no need.

Now equipped with our special impls, we can just go at it:

```rust
bucket_entries(bucket.path())?
	.into_iter()
	.collect::<HashSet<SerializableEntry>>()
	.into_iter()
	.filter_map...
```

And that's it!

**Comment by zkat** [comment]
omg this is an awesome write-up. Thank you!!
